Redis metadata filtering and specification, index customization

Spartee opened the pull request:
### Description

The previous Redis implementation did not allow for the user to specify the index configuration (i.e. changing the underlying algorithm) or add additional metadata to use for querying (i.e. hybrid or "filtered" search).

This PR introduces the ability to specify custom index attributes and metadata attributes as well as use that metadata in filtered queries. Overall, more structure was introduced to the Redis implementation that should allow for easier maintainability moving forward.


### Example data

Suppose we have the following sample data

```python

metadata = [
    {
        "user": "john",
        "age": 18,
        "job": "engineer",
        "credit_score": "high",
    },
    {
        "user": "derrick",
        "age": 14,
        "job": "doctor",
        "credit_score": "low",
    },
    {
        "user": "nancy",
        "age": 94,
        "job": "doctor",
        "credit_score": "high",
    },
    {
        "user": "tyler",
        "age": 100,
        "job": "engineer",
        "credit_score": "high",
    },
    {
        "user": "tim",
        "age": 12,
        "job": "dermatologist",
        "credit_score": "high",
    },
    {
        "user": "taimur",
        "age": 15,
        "job": "CEO",
        "credit_score": "low",
    },
    {
        "user": "joe",
        "age": 35,
        "job": "dentist",
        "credit_score": "medium",
    },
]

texts = ["foo", "foo", "foo", "foo", "bar", "bar", "bar"]

```

# New Features

The following features are now available with the Redis integration into Langchain

## Index schema generation

The schema for the index will now be automatically generated if not specified by the user. For example, the data above has the multiple metadata categories. The the following example

```python

from langchain.embeddings import OpenAIEmbeddings
from langchain.vectorstores.redis import Redis

embeddings = OpenAIEmbeddings()


rds, keys = Redis.from_texts_return_keys(
    texts,
    embeddings,
    metadatas=metadata,
    redis_url="redis://localhost:6379",
    index_name="users"
)
```

Loading the data in through this and the other ``from_documents`` and ``from_texts`` methods will now generate index schema in Redis like the following.

view index schema with the ``redisvl`` tool. [link](redisvl.com)

```bash
$ rvl index info -i users
```


Index Information:
| Index Name   | Storage Type   | Prefixes      | Index Options   |   Indexing |
|--------------|----------------|---------------|-----------------|------------|
| users        | HASH           | ['doc:users'] | []              |          0 |
Index Fields:
| Name           | Attribute      | Type    | Field Option   |   Option Value |
|----------------|----------------|---------|----------------|----------------|
| user           | user           | TEXT    | WEIGHT         |              1 |
| job            | job            | TEXT    | WEIGHT         |              1 |
| credit_score   | credit_score   | TEXT    | WEIGHT         |              1 |
| content        | content        | TEXT    | WEIGHT         |              1 |
| age            | age            | NUMERIC |                |                |
| content_vector | content_vector | VECTOR  |                |                |


### Custom Metadata specification

The metadata schema generation has the following rules
1. All text fields are indexed as text fields.
2. All numeric fields are index as numeric fields.

If you would like to have a text field as a tag field, users can specify overrides like the following for the example data

```python

# this can also be a path to a yaml file
index_schema = {
    "text": [{"name": "user"}, {"name": "job"}],
    "tag": [{"name": "credit_score"}],
    "numeric": [{"name": "age"}],
}

rds, keys = Redis.from_texts_return_keys(
    texts,
    embeddings,
    metadatas=metadata,
    redis_url="redis://localhost:6379",
    index_name="users"
)
```
This will change the index specification to 

Index Information:
| Index Name   | Storage Type   | Prefixes       | Index Options   |   Indexing |
|--------------|----------------|----------------|-----------------|------------|
| users2       | HASH           | ['doc:users2'] | []              |          0 |
Index Fields:
| Name           | Attribute      | Type    | Field Option   | Option Value   |
|----------------|----------------|---------|----------------|----------------|
| user           | user           | TEXT    | WEIGHT         | 1              |
| job            | job            | TEXT    | WEIGHT         | 1              |
| content        | content        | TEXT    | WEIGHT         | 1              |
| credit_score   | credit_score   | TAG     | SEPARATOR      | ,              |
| age            | age            | NUMERIC |                |                |
| content_vector | content_vector | VECTOR  |                |                |


and throw a warning to the user (log output) that the generated schema does not match the specified schema.

```text
index_schema does not match generated schema from metadata.
index_schema: {'text': [{'name': 'user'}, {'name': 'job'}], 'tag': [{'name': 'credit_score'}], 'numeric': [{'name': 'age'}]}
generated_schema: {'text': [{'name': 'user'}, {'name': 'job'}, {'name': 'credit_score'}], 'numeric': [{'name': 'age'}]}
```

As long as this is on purpose,  this is fine.

The schema can be defined as a yaml file or a dictionary

```yaml

text:
  - name: user
  - name: job
tag:
  - name: credit_score
numeric:
  - name: age

```

and you pass in a path like

```python
rds, keys = Redis.from_texts_return_keys(
    texts,
    embeddings,
    metadatas=metadata,
    redis_url="redis://localhost:6379",
    index_name="users3",
    index_schema=Path("sample1.yml").resolve()
)
```

Which will create the same schema as defined in the dictionary example


Index Information:
| Index Name   | Storage Type   | Prefixes       | Index Options   |   Indexing |
|--------------|----------------|----------------|-----------------|------------|
| users3       | HASH           | ['doc:users3'] | []              |          0 |
Index Fields:
| Name           | Attribute      | Type    | Field Option   | Option Value   |
|----------------|----------------|---------|----------------|----------------|
| user           | user           | TEXT    | WEIGHT         | 1              |
| job            | job            | TEXT    | WEIGHT         | 1              |
| content        | content        | TEXT    | WEIGHT         | 1              |
| credit_score   | credit_score   | TAG     | SEPARATOR      | ,              |
| age            | age            | NUMERIC |                |                |
| content_vector | content_vector | VECTOR  |                |                |



### Custom Vector Indexing Schema

Users with large use cases may want to change how they formulate the vector index created by Langchain

To utilize all the features of Redis for vector database use cases like this, you can now do the following to pass in index attribute modifiers like changing the indexing algorithm to HNSW.

```python
vector_schema = {
    "algorithm": "HNSW"
}

rds, keys = Redis.from_texts_return_keys(
    texts,
    embeddings,
    metadatas=metadata,
    redis_url="redis://localhost:6379",
    index_name="users3",
    vector_schema=vector_schema
)

```

A more complex example may look like

```python
vector_schema = {
    "algorithm": "HNSW",
    "ef_construction": 200,
    "ef_runtime": 20
}

rds, keys = Redis.from_texts_return_keys(
    texts,
    embeddings,
    metadatas=metadata,
    redis_url="redis://localhost:6379",
    index_name="users3",
    vector_schema=vector_schema
)
```

All names correspond to the arguments you would set if using Redis-py or RedisVL. (put in doc link later)


### Better Querying

Both vector queries and Range (limit) queries are now available and metadata is returned by default. The outputs are shown.

```python
>>> query = "foo"
>>> results = rds.similarity_search(query, k=1)
>>> print(results)
[Document(page_content='foo', metadata={'user': 'derrick', 'job': 'doctor', 'credit_score': 'low', 'age': '14', 'id': 'doc:users:657a47d7db8b447e88598b83da879b9d', 'score': '7.15255737305e-07'})]

>>> results = rds.similarity_search_with_score(query, k=1, return_metadata=False)
>>> print(results) # no metadata, but with scores
[(Document(page_content='foo', metadata={}), 7.15255737305e-07)]

>>> results = rds.similarity_search_limit_score(query, k=6, score_threshold=0.0001)
>>> print(len(results)) # range query (only above threshold even if k is higher)
4
```

### Custom metadata filtering

A big advantage of Redis in this space is being able to do filtering on data stored alongside the vector itself. With the example above, the following is now possible in langchain. The equivalence operators are overridden to describe a new expression language that mimic that of [redisvl](redisvl.com). This allows for arbitrarily long sequences of filters that resemble SQL commands that can be used directly with vector queries and range queries.

There are two interfaces by which to do so and both are shown. 

```python

>>> from langchain.vectorstores.redis import RedisFilter, RedisNum, RedisText

>>> age_filter = RedisFilter.num("age") > 18
>>> age_filter = RedisNum("age") > 18 # equivalent
>>> results = rds.similarity_search(query, filter=age_filter)
>>> print(len(results))
3

>>> job_filter = RedisFilter.text("job") == "engineer" 
>>> job_filter = RedisText("job") == "engineer" # equivalent
>>> results = rds.similarity_search(query, filter=job_filter)
>>> print(len(results))
2

# fuzzy match text search
>>> job_filter = RedisFilter.text("job") % "eng*"
>>> results = rds.similarity_search(query, filter=job_filter)
>>> print(len(results))
2


# combined filters (AND)
>>> combined = age_filter & job_filter
>>> results = rds.similarity_search(query, filter=combined)
>>> print(len(results))
1

# combined filters (OR)
>>> combined = age_filter | job_filter
>>> results = rds.similarity_search(query, filter=combined)
>>> print(len(results))
4
```

All the above filter results can be checked against the data above.



### TODO

- [x] more tests
- [x] docstrings
- [x] docs

### Other

  - Issue: #3967 
  - Dependencies: No added dependencies
  - Tag maintainer: @hwchase17 @baskaryan @rlancemartin 
  - Twitter handle: @sampartee

Commit:
Redis Meta data Filter changes

Commit:
Add support to Numeric and Tags

Commit:
Remove Print statement

Commit:
Add score_threshold filter to the query

Commit:
Enable maleable index and query formation

Commit:
Restructure

Commit:
fix custom vector schema

Commit:
Re-formatting

tylerhutcherson reviewed (commented):
nice -- left a few comments and questions

Spartee commented on libs/langchain/langchain/vectorstores/redis/redis.py:
```diff
@@ -268,9 +271,15 @@ def add_texts(
+        self,
+        query: str,
+        k: int = 4,
+        meta_filter: Optional[RedisFilter] = None,
```
filter is a python keyword though...

baskaryan replied:
but builtin filter isn't used in this method, so probably fine? (and not like we're expecting folks to subclass this Class and overwrite this method)

baskaryan replied:
also no need for **kwargs if they're not used anywhere

Spartee replied:
I would really press on the fact that this can raise unforeseen errors in the future, but I'll change it to filter. I've been bitten by this kind of thing in larger projects like this.

Spartee commented on libs/langchain/langchain/utilities/redis.py:
```diff
@@ -1,18 +1,81 @@
+    raise ValueError(error_message)
+
+
+def check_index_exists(client: RedisType, index_name: str) -> bool:
```
no preference, but it follows the utility functional pattern

baskaryan commented on libs/langchain/langchain/utilities/redis.py:
```diff
@@ -1,18 +1,81 @@
+        else:
+            self.escaped_chars_re = re.compile(self.DEFAULT_ESCAPED_CHARS)
+
+    def escape(self, value: str) -> str:
```
some unit tests here would be nice

baskaryan commented on libs/langchain/langchain/vectorstores/redis/__init__.py:
```diff
@@ -0,0 +1,4 @@
+from .filters import NumericFilter, TagFilter, TextFilter
```
yea if these are user-facing would prefer to prefix with Redis...

baskaryan commented on libs/langchain/langchain/vectorstores/redis/filters.py:
```diff
@@ -0,0 +1,100 @@
+class RedisFilter:
+    escaper = TokenEscaper()
+
+    def __init__(self, field):
```
linter will yell at you for type hints in all function signatures

baskaryan commented on libs/langchain/langchain/vectorstores/redis/filters.py:
```diff
@@ -0,0 +1,100 @@
+        return self
+
+    def to_string(self) -> str:
+        raise NotImplementedError
```
should this just be abstract?

baskaryan commented on libs/langchain/langchain/vectorstores/redis/redis.py:
```diff
@@ -132,84 +99,113 @@ class Redis(VectorStore):
         index_name: str,
         embedding_function: Callable,
         content_key: str = "content",
-        metadata_key: str = "metadata",
```
maybe worth checking if deprecated params are in kwargs and logging a warning

baskaryan commented on libs/langchain/langchain/vectorstores/redis/redis.py:
```diff
@@ -132,84 +99,113 @@ class Redis(VectorStore):
             raise ValueError(f"Redis failed to connect: {e}")
 
+        # Assume import check is done in get_client
+        from langchain.vectorstores.redis.schema import RedisMetadata
```
nit: import at beginning of method (L123)

baskaryan commented on libs/langchain/langchain/vectorstores/redis/redis.py:
```diff
@@ -308,65 +352,113 @@ def similarity_search_limit_score(
+        )
+        return docs
+
+    def similarity_search_by_vector(
```
nit: parent class already does this

baskaryan commented on libs/langchain/langchain/vectorstores/redis/redis.py:
```diff
@@ -308,65 +352,113 @@ def similarity_search_limit_score(
+        try:
+            from redis.commands.search.query import Query
+        except ImportError:
+            raise ValueError(
```
nit: ImportError

baskaryan commented on libs/langchain/langchain/vectorstores/redis/redis.py:
```diff
@@ -132,84 +99,113 @@ class Redis(VectorStore):
         self.relevance_score_fn = relevance_score_fn
+        self._vector_schema = self._default_content_vector_schema
+        if isinstance(vector_schema, dict):
+            self._vector_schema.update(vector_schema)
```
i believe this will mutate the class attribute (_default_content_vector_schema)
![Screenshot 2023-08-02 at 6 51 14 PM](https://github.com/langchain-ai/langchain/assets/22008038/8e230d78-7874-4802-836f-dd8ec030367b)

Spartee replied:
Great catch!

baskaryan commented on libs/langchain/langchain/vectorstores/redis/redis.py:
```diff
@@ -132,84 +99,113 @@ class Redis(VectorStore):
             return _default_relevance_score
 
+    @property
+    def embeddings(self) -> Optional[Embeddings]:
```
why is this needed again (ik it's from before, just wondering)? for backwards compatibility?

Spartee replied:
Honestly I don't remember. @tylerhutcherson ?

baskaryan commented on libs/langchain/langchain/vectorstores/redis/redis.py:
slight preference to name this file something like `base.py`

Commit:
mypy and address review

Commit:
formatting

baskaryan commented on libs/langchain/langchain/vectorstores/redis/schema.py:
```diff
@@ -0,0 +1,158 @@
+
+
+class TextFieldSchema(BaseField):
+    weight: Optional[float] = 1
```
out of curiosity why are all of these Optional if they have non-null defaults (other than phoenetic_matcher)? is that based on TextField

baskaryan commented on libs/langchain/langchain/vectorstores/redis/redis.py:
```diff
@@ -132,84 +110,128 @@ class Redis(VectorStore):
-        # TODO: Accept embedding object directly
-        return None
+        # set default vector schema and update with user provided schema
+        self._vector_schema = {}
+        self._vector_schema.update(self._default_content_vector_schema)
+        if isinstance(vector_schema, dict):
+            self._vector_schema.update(vector_schema)
```
```suggestion
vector_schema = vector_schema or {}
self._vector_schema = {**self._default_content, **vector_schema}
```

baskaryan commented on libs/langchain/langchain/vectorstores/redis/redis.py:
```diff
@@ -132,84 +110,128 @@ class Redis(VectorStore):
+
+        # set vector distance metric and scoring function
+        self._distance_metric = str(self._vector_schema["distance_metric"]).upper()
+        self._select_relevance_score_fn()
```
this doesn't set anything, why call here?

baskaryan commented on libs/langchain/langchain/vectorstores/redis/schema.py:
```diff
@@ -0,0 +1,158 @@
+    from redis.commands.search.field import Field as RedisField
+
+
+class BaseField(BaseModel):
```
most of the FieldSchemas seem identical to the redis Fields they're wrapping — what's the function of these again?

baskaryan commented on libs/langchain/langchain/vectorstores/redis/redis.py:
```diff
@@ -308,65 +380,106 @@ def similarity_search_limit_score(
+    ) -> List[Tuple[Document, float]]:
+        """Run similarity search with distance."""
+        docs = self._similarity_search(query, k=k, filter=filter, return_scores=True)
+        return docs  # type: ignore
```
to avoid having to ignore mypy, could just move all of _similarity_search into similarity_search_with_score, return docs and scores, and then similarity_search looks like
```python
docs_and_scores = self.similarity_search_with_scores(...)
return [d for d, _ in docs_and_scores]
```

Spartee replied:
So introduce a loop just to satisfy mypy? This and code duplication was my issue with the last approach.

baskaryan commented on libs/langchain/langchain/vectorstores/redis/redis.py:
```diff
@@ -243,6 +265,13 @@ def add_texts(
+        if metadatas:
+            if isinstance(metadatas, list) and len(metadatas) != len(texts):  # type: ignore
+                raise ValueError("Number of metadatas must match number of texts")
+            if not isinstance(metadatas, list) and not isinstance(metadatas[0], dict):
```
sounds like this should be an or (if either condition isn't met, throw error?)

baskaryan commented on libs/langchain/langchain/vectorstores/redis/redis.py:
```diff
@@ -281,11 +316,47 @@ def similarity_search(
+        if score_threshold is None:
+            redis_query = self._prepare_vector_query(k, filter=filter)
+        else:
+            redis_query = self._prepare_range_query(k, filter=filter)
```
is redis_query supposed to be identical in both if branches?

Spartee replied:
yes.

baskaryan reviewed (commented):
would be good to have someone from redis side review the redis-specific logic!

Commit:
Refactor redis vector store

Commit:
Add tests for new capability

Commit:
remove prints

baskaryan commented on libs/langchain/langchain/vectorstores/redis/base.py:
```diff
@@ -243,6 +306,13 @@ def add_texts(
+        if metadatas:
+            if isinstance(metadatas, list) and len(metadatas) != len(texts):  # type: ignore
+                raise ValueError("Number of metadatas must match number of texts")
+            if not isinstance(metadatas, list) and not isinstance(metadatas[0], dict):
```
```suggestion
            if not (isinstance(metadatas, list) and isinstance(metadatas[0], dict)):
```

baskaryan replied:
currently this would pass for eg a tuple of dicts or a list of non-dicts

baskaryan commented on libs/langchain/langchain/vectorstores/redis/base.py:
```diff
@@ -132,84 +144,135 @@ class Redis(VectorStore):
         self,
         redis_url: str,
         index_name: str,
         embedding_function: Callable,
```
since we're already making breaking changes, i'd suggest we update this to be
```suggestion
        embedding: Embeddings,
```
which is interface all newer VectorStores have

baskaryan commented on libs/langchain/langchain/vectorstores/redis/base.py:
```diff
@@ -281,11 +358,62 @@ def similarity_search(
+        if return_scores:
+            return_fields.append("score")
+
+        if score_threshold is None:
+            redis_query = self._prepare_vector_query(
+                k, filter=filter, return_fields=return_fields
+            )
+        else:
+            redis_query = self._prepare_range_query(
+                k, filter=filter, return_fields=return_fields
+            )
+            params_dict["score_threshold"] = str(score_threshold)
```
```suggestion
        redis_query = self._prepare_vector_query(
            k, filter=filter, return_fields=return_fields
        )
        if score_threshold is not None:
            params_dict["score_threshold"] = str(score_threshold)
```

Spartee replied:
I might be missing something here, but if there is a score threshold, it needs to call _prepare_range_query()

...this wouldn't work? (might be github viewer tbh)

baskaryan commented on libs/langchain/langchain/vectorstores/redis/base.py:
```diff
@@ -281,11 +358,62 @@ def similarity_search(
+            if return_metadata:
+                metadata = {k: getattr(result, k) for k in self._schema.metadata_keys}
+                metadata["id"] = result.id
+                metadata["score"] = result.score
```
im not sure we want to include score in metadata. a document's contents and metadata should be independent of the particular query that caused them to be returned

Spartee replied:
I debated this. but I'm ok with that.

Commit:
interim commit

Commit:
interim commit

Commit:
Add metadata cleaning, docstrings, docs

Add the ability to clean the metadata before
it goes into redis enabling document_loaders
that return lists of strings to create categorical
values for Tags in Redis indices.

Also, added docstrings and updated the jupyter notebook

Commit:
Merge branch 'master' into redis-refactor

Commit:
Fix some linting errors

baskaryan committed:
lint

baskaryan commented:
fixed some of the lint issues here #9705 if you want to merge that into this pr

tylerhutcherson commented on libs/langchain/langchain/vectorstores/redis/base.py:
```diff
@@ -0,0 +1,1199 @@
+        base_query = f"({query_prefix})=>[KNN {k} @{vector_key} $vector AS score]"
+
+        query = (
+            Query(base_query).return_fields(*return_fields).sort_by("score").dialect(2)
```
still need to add `.paging(0, k)` here
```suggestion
            Query(base_query)
            .return_fields(*return_fields)
            .sort_by("score")
            .paging(0, k)
            .dialect(2)
```

alonre24 commented on libs/langchain/langchain/vectorstores/redis/base.py:
```diff
@@ -0,0 +1,1199 @@
+                from langchain.vectorstores import Redis
+                from langchain.embeddings import OpenAIEmbeddings
+                embeddings = OpenAIEmbeddings()
+                redisearch, keys = RediSearch.from_texts_return_keys(
```
Isn't it `Redis.from_texts_return_keys` ?

Commit:
Address from_existing issues

Commit:
minor context_key bug fix

Commit:
Fix semantic cache

alonre24 commented on libs/langchain/langchain/vectorstores/redis/filters.py:
```diff
@@ -0,0 +1,429 @@
+    OPERATOR_MAP = {
+        RedisFilterOperator.EQ: '@%s:"%s"',
+        RedisFilterOperator.NE: '(-@%s:"%s")',
+        RedisFilterOperator.LIKE: "@%s:%s",
```
I'm not sure that LIKE represents what you meant it would be. The difference between `@%s:"%s"` and `@%s:%s` is that the first one allows you to find exact matches **for phrases**, while the last one also looks for exact matches but for each token separately. Also, using `@%s:%s` (without quotes) will support [stemming](https://redis.io/docs/interact/search-and-query/advanced-concepts/stemming/) by default.
To get a behavior that is more similar to SQL's LIKE operator, we can use prefix, infix or suffix matching (see [query](https://redis.io/docs/interact/search-and-query/query/) docs).

alonre24 reviewed (commented):
Added a few RediSearch syntax-related comments

alonre24 commented on libs/langchain/langchain/vectorstores/redis/base.py:
```diff
@@ -0,0 +1,1351 @@
+            Query(query_string)
+            .return_fields(*return_fields)
+            .sort_by("distance")
+            .paging(0, k)
```
One of the advantages of range queries is that you can get all the results that are within a given range (not just the top k ones), and in particular you can see how many results are within the range. To utilize this feature I would suggest having k as an optional argument (and perhaps using an upper bound such as 1000 for paging), if that makes sense to you.

tylerhutcherson replied:
agreed -- was going to suggest using K as an optional upper bound. @alonre24 isn't paging a default param 0-10 though?

Maybe something like:
```python
query = (
    Query(query_string)
    .return_fields(*return_fields)
    .sort_by("distance")
    .dialect(2)
)

if k:
    query = query.paging(0, k)
```

alonre24 replied:
Yes, the default paging is 10, that's why I suggest also choosing an upper bound, so we have: 
```py
query = query.paging(0, k if k else UPPER_BOUND)
```

Spartee replied:
It's part of the higher level interface abstraction so it would have to be set. I've started this conversation the lc folks though. Most likely the best route is another method that exposes this feature better.

alonre24 commented on libs/langchain/langchain/vectorstores/redis/base.py:
```diff
@@ -0,0 +1,1351 @@
+
+        # if it's a list of strings, we assume it's a tag
+        if isinstance(value, (list, tuple)):
+            if not value or isinstance(value[0], str):
```
why `if not value` is ok?

Spartee replied:
same thing as saying ``len(value) == 0``. 

```python 
>>> x = []
>>> if not x:
...     print("it's like an emptiness check")
... 
it's like an emptiness check
```

looks cleaner than len check or try/except

alonre24 commented on libs/langchain/langchain/vectorstores/redis/base.py:
```diff
@@ -0,0 +1,1351 @@
+        # if it's a list/tuple of strings, we join it
+        elif isinstance(value, (list, tuple)):
+            if not value or isinstance(value[0], str):
+                clean_meta[key] = ",".join(value)
```
Note that if there are tag values within the list that contain a comma, they will be split when indexed in RediSearch. Consider validating that there are no such values, or allowing the user to specify a different separator (API allows it).

Spartee replied:
So currently, they could specify it in the schema and clean up the data themselves beforehand, but for the automatically generated metadata (how most will use it because they just use the data loaders), it's defaulting to ``,`` right now. Been thinking I should have a default here anyway. Good call out.

alonre24 commented on libs/langchain/langchain/vectorstores/redis/filters.py:
```diff
@@ -0,0 +1,429 @@
+
+        >>> from langchain.vectorstores.redis import RedisTag, RedisNum
+        >>> brand_is_nike = RedisTag("brand") == "nike"
+        >>> price_is_over_100 = RedisNum("price") < 100
```
```suggestion
        >>> price_is_under_100 = RedisNum("price") < 100
```

Commit:
Address Embeddings interface and retriever tests

Commit:
Merge branch 'fix-semantic-cache' into redis-refactor

Commit:
interim commit for cache fix

Commit:
interim commit for cache fix (dirty)

Commit:
Cache tests passing with real embeddings

baskaryan committed:
lint

Commit:
Merge remote-tracking branch 'upstream/bagatur/redis_refactor' into redis-refactor

Commit:
linting fixes

meiravgri commented on libs/langchain/langchain/vectorstores/redis/schema.py:
```diff
@@ -0,0 +1,276 @@
+
+    # filled by default_vector_schema
+    vector: Optional[List[Union[FlatVectorField, HNSWVectorField]]] = None
+    content_key: str = "content"
```
If I understand correctly, having a field called "content" is mandatory.
Fix me if I'm wrong but the only location I found it might be a problem to give this field a user-defined name is in `similarity_search_with_score` where we return `result.content` 
(in `similarity_search()` it is handled better IMO by using `getattr(result, content_key)` instead)
Another solution is to define the return fields as
`Query().return_field(self.content_field, as_field="content")`

Also, where is the field name enforced in `from_existing_index()`?

meiravgri commented on libs/langchain/tests/integration_tests/vectorstores/test_redis.py:
```diff
@@ -40,30 +72,55 @@ def test_redis_new_vector(texts: List[str]) -> None:
 def test_redis_from_existing(texts: List[str]) -> None:
     """Test adding a new document"""
-    Redis.from_texts(
+    docsearch = Redis.from_texts(
```
I'd add tests that create or connect to an existing index that doesn't include the default fields names

meiravgri reviewed (commented):
I know it was already merged but I had a few comments :)
